software/contrib/consequencer.py

```diff
@@ -211,88 +221,200 @@ def clockTriggerEnd():
 
     def generateRandomPattern(self, length, min, max):
-        self.t=[]
+        self.t = []
         for i in range(0, length):
-            self.t.append(uniform(0,9))
+            self.t.append(uniform(0, 9))
         return self.t
 
-
     def getRandomness(self):
         # If mode 1 and there is CV on the analogue input use it, if not use the knob position
-        val = 100 * ain.percent()
-        if self.analogInputMode == MODE_RANDOM and val > self.minAnalogInputVoltage:
-            self.randomness = min(val + k1.read_position(), 99)
+        if self.analogInputMode == MODE_RANDOM and self.ainVal > self.minAnalogInputVoltage:
+            self.randomness = min(self.ainVal + self.k1Val, 99)
         else:
-            self.randomness = k1.read_position()
+            self.randomness = self.k1Val
+
+        if self.randomness_prev != self.randomness:
+            self.randomness_prev = self.randomness
+
+            if not self.screenOff:
+                self._updateUI = True
+
+    def getAinVal(self):
+        # Read ain val and update if > threshold
+        self.ainValTemp = 100 * ain.percent()
+        if abs(self.ainValTemp - self.ainVal) > AIN_CHANGE_TOLERANCE:
+            self.ainVal = self.ainValTemp
+
+
+    def getKnobVals(self):
+        # Read knob vals and update if > threshold
+        self.k1ValTemp = k1.read_position()
+        if abs(self.k1ValTemp - self.k1Val) > KNOB_CHANGE_TOLERANCE:
+            self.k1Val = self.k1ValTemp
+            self.screenOff = False
+            self.lastInteractionTimeMs = ticks_ms()
+
+        self.k2ValTemp = k2.read_position(len(self.BD))
+        if abs(self.k2ValTemp - self.k2Val) > KNOB_CHANGE_TOLERANCE:
+            self.k2Val = self.k2ValTemp
+            self.screenOff = False
+            self.lastInteractionTimeMs = ticks_ms()
 
     def main(self):
         while True:
+            self.getAinVal()
+            self.getKnobVals()
             self.getPattern()
             self.getRandomness()
             self.getCvPattern()
-            self.updateScreen()
+            # Update screen if updateUI flag has been set
+            if self._updateUI:
+                self.updateScreen()
+                self._updateUI = False
             # If I have been running, then stopped for longer than reset_timeout, reset the steps and clock_step to 0
-            if self.clock_step != 0 and ticks_diff(ticks_ms(), din.last_triggered()) > self.reset_timeout:
+            if (
+                self.clock_step != 0
+                and ticks_diff(ticks_ms(), din.last_triggered()) > self.reset_timeout
+            ):
                 self.step = 0
                 self.clock_step = 0
 
+            # Has the module been left along for a while, turn off the screen
+            if (
+                not self.screenOff
+                and ticks_diff(ticks_ms(), self.lastInteractionTimeMs) > self.screenOffTimeoutMs
+            ):
+                self.drawBlankScreen()
```

Comment:
Possible place to use @chrisib's screensaver class, not a requirement for this PR but in future might reduce the amount of code needed to implement a new screensaver

Reply:
I did indeed get inspiration for the screen saver from @chrisib . However, I chose to implement a simpler version to meet my more minimal requirements.

Reply:
Sweet, sounds good to me!